File: mobile/lib/network/pairing.dart
import 'package:shared_preferences/shared_preferences.dart';

import 'wheeldeck_client.dart';

/// Persists the session token issued after a successful pairing.
abstract class SessionTokenStore {
  Future<String?> load();
  Future<void> save(String token);
}

/// Stores the token in the platform's shared preferences. Works across the
/// native targets and the iOS PWA (shared_preferences uses localStorage on web).
class SharedPreferencesSessionTokenStore implements SessionTokenStore {
  static const String _key = 'wheeldeck.session_token';

  @override
  Future<String?> load() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_key);
  }

  @override
  Future<void> save(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_key, token);
  }
}

/// Drives the pairing flow: submits the entered or scanned code, persists the
/// issued session token, and restores it on later sessions so re-pairing can be
/// skipped.
class PairingController {
  PairingController({
    required this.store,
    required this.client,
  }) {
    client.onPairingAccepted(store.save);
  }

  final SessionTokenStore store;
  final WheelDeckClient client;

  /// Loads the session token persisted from a previous session, if any.
  Future<String?> restoreSession() => store.load();

  /// Sends the pairing code to the desktop for validation.
  void submitPairingCode(String code) => client.submitPairingCode(code);
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/services/Pairing.kt
package dev.fazrigading.wheeldeck.data.services

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

private val Context.sessionTokenDataStore by preferencesDataStore(name = "wheeldeck_session")

/// Persists the session token issued after a successful pairing.
interface SessionTokenStore {
    suspend fun load(): String?
    suspend fun save(token: String)
}

/// Stores the token in DataStore Preferences.
class DataStoreSessionTokenStore(context: Context) : SessionTokenStore {
    private val dataStore = context.applicationContext.sessionTokenDataStore

    override suspend fun load(): String? =
        dataStore.data.first()[stringPreferencesKey(KEY)]

    override suspend fun save(token: String) {
        dataStore.edit { it[stringPreferencesKey(KEY)] = token }
    }

    companion object {
        const val KEY = "wheeldeck.session_token"
    }
}

/// Drives the pairing flow: submits the entered or scanned code, persists the
/// issued session token, and restores it on later sessions so re-pairing can be
/// skipped.
class PairingController(
    private val store: SessionTokenStore,
    private val client: WheelDeckClient,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    init {
        client.onPairingAccepted { token -> scope.launch { store.save(token) } }
    }

    /// Loads the session token persisted from a previous session and restores it
    /// onto the client so the next connection can skip pairing.
    suspend fun restoreSession(): String? {
        val token = store.load()
        if (token != null) {
            client.setSessionToken(token)
        }
        return token
    }

    /// Sends the pairing code to the desktop for validation.
    fun submitPairingCode(code: String) = client.submitPairingCode(code)
}
